### PR TITLE
41/integrate edit section title/description

### PR DESCRIPTION
**Pull request by hamishtaplin**
### What is the context of this PR?

[User story](https://trello.com/c/MpEfpwSH/41-edit-section-with-title-and-description-5)

- "Create a questionnaire" user journey now creates an empty section (via API)
- `WithQuestionnaire` HoC wraps any components that read/write to `questionnaire` part of api
- Wired section UI up to the store
- Simplified how the `BaseLayout`, `Breadcrumb` and `Nav` components decide how/what to render by passing `questionnaire` data around
- Refactored all code to use `withChangeHandler` HoC to manage dispatching of form change events to their subscribers.
- Cleaned up and fixed some layout issues between existing "pages" 

### How to review 

- create/edit a questionnaire, ensuring that section title and description are saved to API

### CODE REVIEW

**Comment by hamishtaplin** [comment]
[Edit Section with Title and Description (5)](https://trello.com/c/MpEfpwSH/41-edit-section-with-title-and-description-5)

File: src/components/WithQuestionnaire/WithQuestionnaire.test.js [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+  it("should catch and prevent form submit events", () => {
+    const preventDefault = jest.fn();
+    wrapper.simulate("submit", { preventDefault });
+    expect(preventDefault).toHaveBeenCalled();
+  });
+
+  it("should store updated values in state", () => {
+    const value = { name: "questionnaire.title", value: "My Title" };
+    wrapper.instance().handleChange(value);
```
Comment (WickyNilliams) [inline src/components/WithQuestionnaire/WithQuestionnaire.test.js]: why not `simulate("change")`?

File: src/components/WithQuestionnaire/WithQuestionnaire.test.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    expect(wrapper.state().questionnaire).toEqual({ title: "My Title" });
+  });
+
+  it("should save to API on blur event", () => {
+    const value = { name: "questionnaire.title", value: "My Title" };
+    wrapper.instance().handleChange(value);
+    wrapper.instance().handleBlur();
```
Comment (WickyNilliams) [inline src/components/WithQuestionnaire/WithQuestionnaire.test.js]: same here, why not simulate?

File: src/containers/QuestionnaireMetaPage/index.js [hunk 3]
```diff
@@ -6,7 +6,7 @@ import updateQuestionnaire from "queries/updateQuestionnaire";
 import QuestionnaireMeta from "./QuestionnaireMeta";
 
 export const mapStateToProps = (state, ownProps) => ({
-  id: ownProps.match.params.id
+  questionnaireId: ownProps.match.params.questionnaireId
 });
 
 export const mapResultsToProps = ({ data, ownProps }) => {
```
Comment (WickyNilliams) [inline src/containers/QuestionnaireMetaPage/index.js]: again could use lodash pick here:

```js
export const mapResultsToProps = ({ data }) => pick(data, ["questionnaire", "loading"])
```

File: src/containers/QuestionnaireMetaPage/index.js [hunk 4]
```diff
@@ -20,13 +20,13 @@ export const mapResultsToProps = ({ data, ownProps }) => {
 
 export const withData = graphql(getQuestionnaire, {
   props: mapResultsToProps,
-  options: props => ({ variables: { id: props.id } })
+  options: props => ({ variables: { id: props.questionnaireId } })
 });
 
 export const withMutation = graphql(updateQuestionnaire, {
   props: ({ ownProps, mutate }) => ({
```
Comment (WickyNilliams) [inline src/containers/QuestionnaireMetaPage/index.js]: ownProps unused (surprised linter didn't kick off about this?)

File: src/containers/QuestionnaireDesignPage/QuestionnaireDesign.js [hunk 5]
```diff
@@ -45,22 +49,18 @@ export class QuestionnaireDesignPage extends React.Component {
     if (sectionId !== this.state.focused) {
       this.setState({ focused: sectionId });
     }
+
+    this.props.update(this.state);
```
Comment (WickyNilliams) [inline src/containers/QuestionnaireDesignPage/QuestionnaireDesign.js]: This causes a mutation on blur and focus, so tabbing from one field to another sends two mutations

File: src/containers/QuestionnaireDesignPage/QuestionnaireDesign.js [hunk 6]
```diff
@@ -1,40 +1,44 @@
...
-import { set, noop } from "lodash";
 
-export class QuestionnaireDesignPage extends React.Component {
+export class QuestionnaireDesignPage extends Component {
   static propTypes = {
-    loading: PropTypes.bool.isRequired,
-    questionnaire: CustomPropTypes.questionnaire
+    breadcrumb: CustomPropTypes.breadcrumb,
+    onSubmit: PropTypes.func.isRequired,
+    update: PropTypes.func.isRequired,
```
Comment (hamishtaplin) [inline src/containers/QuestionnaireDesignPage/QuestionnaireDesign.js]: doesn't that imply there's some event handling happening?

File: src/containers/QuestionnaireDesignPage/index.js [hunk 7]
```diff
@@ -1,21 +1,58 @@
...
+  const { sectionId } = ownProps;
+
+  let props = {
+    questionnaire,
+    loading
+  };
+
+  if (questionnaire) {
+    const section = find(questionnaire.sections, {
+      id: parseInt(sectionId, 10)
```
Comment (hamishtaplin) [inline src/containers/QuestionnaireDesignPage/index.js]: Yeah it's already caused some issues. What needs doing?

**Pull request merged**